### ISSUE TITLE
Predefine errors returned by the library code

### ISSUE THREAD

**Issue by mbwalas** [google/goacme#15]
In order to avoid code like:
`strings.Contains(err.Error(), "Registration key is already in use")`
every error returned by the library should be predefined as var, see for example https://golang.org/pkg/net/http/#pkg-variables

**Comment by crhym3** [comment]
ACME spec has lots of errors, many more than net/http.
What if we had something like this instead:

```go
type Error struct {
  Code string
  Message string
}

func (e *Error) Error() string { ... }
```

And, of course, we can expose all error codes ACME defines as constants.
Otherwise, I don't mind creating error vars either.

Let me check the spec to verify how many errors there are exactly...

**Comment by mbwalas** [comment]
If we can't get all as consts (ideal) then we can instead have "kinds" of errors that are reasonable for clients to act upon https://letsencrypt.github.io/acme-spec/#rfc.section.5.4

Where do you see more errors defined than net/http ?

**Comment by crhym3** [comment]
Wait, I didn't realize we already have `goacme.Error` type!

Anyway, coming back to the errors, given that error table posted in the spec, how would this acme client identify `Registration key is already in use`? It's not in the table.

**Comment by mbwalas** [comment]
Yes, I do agree the protocol is a bit broken here. Maybe there is some HTTP code in that case that let us derive this situation. 

I agree that to start with this different problems: https://github.com/letsencrypt/boulder/blob/master/probs/probs.go

when we look at the usage: https://github.com/letsencrypt/boulder/blob/master/wfe/web-front-end.go#L492

if I would be able to say `goacme.IsConflict(err)` then it would be enough in context of `new-reg` to render registration key already in use flow. I am not entirely sure if what I proposed is the right method (golang wise) for distinguishing between errors.

**Comment by crhym3** [comment]
Right. I see, they're actually checking account existence by making additional round trip before `new-reg`. If feels like a caller of `client.Register` would be a better fit for this additional functionality, wouldn't you agree?

The reason I haven't predefined the errors is because they have two fields: `type` and `detail`. The `type` is predefined in that table you linked to, but detail is not, so it can't be a constant.

Of course, we could predefine those errors nonetheless and use that same table to populate the `detail` field, but then callers of the client will loose real error response messages from a CA server.

Anyway, I have an idea. Let me create a pull request to show what I mean.

**Comment by crhym3** [comment]
Fixed by #21
